dali/python/nvidia/dali/worker.py

```diff
@@ -0,0 +1,221 @@
+        self.batch_i = 0
+
+    def _next_batch_i(self):
+        res = self.batch_i
+        self.batch_i = (self.batch_i + 1) % len(self.mem_chunks)
+        return res
+
+    def next_mem_chunk(self):
+        batch_i = self._next_batch_i()
+        return self.mem_chunks[batch_i]
+
+    def close(self):
+        for chunk in self.mem_chunks:
+            chunk.close()
+
+# TODO(klecki): Add exception handling to this thread
+def dispatcher(batch_dispatcher, ready_cv, ready_queue):
+    """Receives batches produced in the main thread and dispatches them to the parent process.
+    It is run in a separate thread because both callback and dispatcher may
+    wait on IO operations a lot and in that case Python threads provide some performance gain.
+    """
+    while True:
+        with ready_cv:
+            while len(ready_queue) == 0:
+                ready_cv.wait()
+            message = ready_queue.pop(0)
+        if message is None:
+            break
+        batch_dispatcher.send(message)
+
+
+def receiver(task_pipe, tasks_cv, tasks_queue):
+    """Receives list of tasks scheduled to be done by the worker, run in a separate thread
+    to avoid blocking of the main process when it schedules another batch for the worker and
+    worker is busy with previously scheduled computations.
+    """
+    try:
+        while True:
+            scheduled = task_pipe.recv()
+            if scheduled is None:
+                break
+            with tasks_cv:
+                tasks_queue.append(scheduled)
+                tasks_cv.notify()
+    finally:
+        with tasks_cv:
+            tasks_queue.insert(0, None)
+            tasks_cv.notify()
+
+
+def worker(worker_id, callbacks, prefetch_queue_depths, initial_chunk_size, task_pipe, res_pipe, sock):
+    """Entry point of worker process.
+
+    Computes the data in the main thread, in separate threads:
+    * waits for incoming tasks,
+    * serializes results and passes them to the main process.
+
+    Parameters
+    ----------
+    `callbacks` : callable list
+        List of callables that worker can call to perform a (part of parallelized) task.
+    `prefetch_queue_depths` : int
+        Number of shared memory chunks that should be allocated per callaback, used in cycle buffer manner
+        to pass callback results to parent process.
+    `initial_chunk_size` : int
+        Initial size of shared memory chunk.
+    `task_pipe`: Pipe
+        Pipe used to read list of tasks that given callback should be run on to produce (part of a) result batch.
+    `res_pipe`: Pipe
+        Pipe used to notify the parent process about another batch ready to read in the given memory chunk.
+    `sock` : socket
+        Python wrapper around Unix socket used to pass file descriptors identifying shared memory chunk to parent process.
+    """
+    contexts = None
+    ready_cv = threading.Condition()
+    tasks_cv = threading.Condition()
+    ready_queue, tasks_queue = [], []
```

Comment:
Why not use python `Queue` directly?

Reply:
Conscious decision to control the threading on our own, I think I heard that Queue introduces its own threads. But maybe it was the one from multiprocess.

Reply:
Mhh, I would be very surprise seeing Queue introduces its own thread. I would rather expect it to encapsulate cv and waiting pattern.

Reply:
It does introduce its thread
>When a process first puts an item on the queue a feeder thread is started which transfers objects from a buffer into the pipe.
https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Queue

There is SimpleQueue that doesn't introduce its own thread and "task-tracking" semantics, but in docs it says
>New in version 3.7.

Reply:
😢 
Let it be.